core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java

```diff
@@ -513,9 +519,30 @@ private RehashIterator(Iterable<IntermediateOperation> intermediateOperations) {
       protected S getNext() {
          do {
             if (currentIterator == null) {
-               log.tracef("Creating rehash iterator for segments %s", segmentsToUse);
-               currentIterator = nonRehashRemoteIterator(segmentsToUse, receivedKeys::get,
-                     publisherDecorator(completedHandler, lostSegments -> {
+               // This covers when an iterator invokes hasNext multiple times each time returning false
+               // Otherwise the second hasNext would block for the timeout since we may not ever see a topology update
+               if (complete) {
```

Comment:
`complete` is set after the loop, so I think this check should be before the loop.

The `if (currentIterator == null)` body is also getting a little large, so I'd start the loop with the "else" branch instead:

            if (currentIterator != null && currentIterator.hasNext()) {
               return currentIterator.next();
            }

If this doesn't return you know the currentIterator is either null or exhausted, so you don't need to set `currentIterator = null`.

Reply:
I have to play around with this a bit, as it isn't this straight forward unfortunately.

Reply:
I can't really move that code up higher, as we shouldn't check the cache topology unless both the iterator is exhausted and segmentsToUse is also not empty. It sadly isn't just a simple else.

Also the completed adds an additional variable access per entry, which by keeping it inside the if prevents. But I can still change that if you really wanted.

Reply:
It took a while to understand what you meant with segmentsToUse also not being empty, but if I got it right this time then I don't think you need the complete field at all:

         while (true){
            if (currentIterator != null && currentIterator.hasNext()) {
               return currentIterator.next();
            }

            // Either we don't have an iterator or the current iterator is exhausted
            if (segmentsToUse.isEmpty()) {
               // No more segments to spawn new iterators
               return null;
            }

            // An iterator completes all segments, unless we either had a node leave (SuspectException)
            // or a new node came up and data rehashed away from the node we requested from.
            // In either case we need to wait for a new topology before spawning a new iterator.
            if (currentIterator != null) {
               try {
                  int nextTopology = cacheTopology.getTopologyId() + 1;
                  log.tracef("Waiting for topology %d to continue iterator operation with segments %s", nextTopology,
                        segmentsToUse);
                  stateTransferLock.topologyFuture(nextTopology).get(timeout, timeoutUnit);
               } catch (InterruptedException | ExecutionException | TimeoutException e) {
                  throw new CacheException(e);
               }
            }

            cacheTopology = dm.getCacheTopology();
            log.tracef("Creating non-rehash iterator for segments %s using topology id: %d", segmentsToUse, cacheTopology.getTopologyId());
            currentIterator = nonRehashRemoteIterator(cacheTopology.getReadConsistentHash(), segmentsToUse,
                  receivedKeys::get, publisherDecorator(completedHandler, lostSegments -> {
                  }, k -> {
                     // Every time a key is retrieved from iterator we add it to the keys received
                     // Then when we retry we exclude those keys to keep out duplicates
                     Set<Object> set = receivedKeys.get(keyPartitioner.getSegment(k));
                     if (set != null) {
                        set.add(k);
                     }
                  }), intermediateOperations);
         }
      }

Reply:
Yeah tbh this is more what I wanted to do, but I was quite distracted yesterday. So thanks for this :) I will try it out.